internal/blocksync/reactor.go

```diff
@@ -284,6 +289,8 @@ func (bcR *Reactor) Receive(e p2p.Envelope) {
 // Handle messages from the poolReactor telling the reactor what to do.
 // NOTE: Don't sleep in the FOR_LOOP or otherwise slow it down!
 func (bcR *Reactor) poolRoutine(stateSynced bool) {
+	defer bcR.poolRoutineWg.Done()
```

Comment:
Honestly, the `bcR.poolRoutineWg.Add(1)` should come here, instead of putting whenever this routine is started. We add a permit to the semaphore when starting this routine, then we defer returning the permit.

Reply:
I might be misunderstanding the comment here.  The pattern with a waitgroup is to `Add` before (in the caller goroutine) launching it with `go` and `defer` the `Done` in the goroutine.  e.g.

```
wg.Add(1)
go func() {
	defer wg.Done()
	// .. do stuff e.g. bcR.poolRoutine(true)
	time.Sleep(time.Second)
}()

wg.Wait()
```

Such as in the cometbft example: https://github.com/cometbft/cometbft/blob/8acc13c31c86fc3dff9da071a1a151b1c5ca46e5/cmd/cometbft/commands/compact.go#L46-L48

Reply:
You are probably right here, we should first add the permit to then start the thread, otherwise the thread may not be scheduled soon enough to add the permit.